File: src/application/services/Product.service.ts
import { Inject, Injectable } from "@nestjs/common";
import { EntityManager } from "typeorm";

import {
  Product,
  Product_Catalog,
  Product_Category,
  Product_Detail,
  Product_Image,
  Product_Option_Group,
  Product_Price,
  Product_Summary,
  Product_Tag,
} from "src/domain/entities";
import IRepository from "src/domain/repositories/IRepository";
import { ProductRepository } from "src/infrastructure/repositories";
import { FilterDTO, ProductInputDTO } from "../dto";

@Injectable()
export default class ProductService {
  constructor(
    private readonly entity_manager: EntityManager,
    @Inject("IProductRepository")
    private readonly repository: IRepository<Product | Product_Summary | Product_Catalog>,
    @Inject("IProductDetailRepository")
    private readonly product_detail_repository: IRepository<Product_Detail>,
    @Inject("IProductPriceRepository")
    private readonly product_price_repository: IRepository<Product_Price>,
    @Inject("IProductCategoryRepository")
    private readonly product_category_repository: IRepository<Product_Category>,
    @Inject("IProductOptionGroupRepository")
    private readonly product_option_group_repository: IRepository<Product_Option_Group>,
    @Inject("IProductImageRepository")
    private readonly product_image_repository: IRepository<Product_Image>,
    @Inject("IProductTagRepository")
    private readonly product_tag_repository: IRepository<Product_Tag>,
  ) {}

  async register({
    detail,
    price,
    categories,
    option_groups,
    images,
    tags: tag_ids,
    seller_id,
    brand_id,
    ...product
  }: ProductInputDTO) {
    try {
      // 상품 등록 트랜잭션 처리
      const product_entity = await this.entity_manager.transaction(async (manager) => {
        // 상품 등록
        const product_entity = await new ProductRepository(manager).save({
          ...product,
          seller_id,
          brand_id,
        });
        const { id: product_id } = product_entity;

        // 상품 상세 등록
        await this.product_detail_repository.with_transaction(manager).save({
          ...detail,
          product_id,
        });

        // 상품 가격 등록
        await this.product_price_repository.with_transaction(manager).save({
          ...price,
          product_id,
        });

        // 상품 카테고리 등록
        await this.product_category_repository
          .with_transaction(manager)
          .saves(categories.map((category) => ({ ...category, product_id })));

        // 상품 옵션 등록
        await this.product_option_group_repository
          .with_transaction(manager)
          .saves(option_groups.map((group) => ({ ...group, product_id })));

        // 상품 이미지 등록
        await this.product_image_repository
          .with_transaction(manager)
          .saves(images.map((image) => ({ ...image, product_id })));

        // 상품 태그 등록
        await this.product_tag_repository
          .with_transaction(manager)
          .saves(tag_ids.map((tag_id) => ({ tag_id, product_id })));

        return product_entity;
      });
      // 상품 등록 결과 반환
      return (({ id, name, slug, created_at, updated_at }) => ({
        id,
        name,
        slug,
        created_at,
        updated_at,
      }))(product_entity);
    } catch (error) {
      throw new Error((error as Error).message);
    }
  }

  async find_all({ page = 1, per_page = 10, sort, ...rest }: FilterDTO) {
    const [sort_field, sort_order] = sort?.split(":") ?? ["created_at", "DESC"];

    const items = await this.repository.find_by_filters({
      page,
      per_page,
      sort_field,
      sort_order,
      ...rest,
    });

    // 페이지네이션 요약 정보
    const pagination = {
      total_items: items.length,
      total_pages: Math.ceil(items.length / (per_page ?? 10)),
      current_page: page ?? 1,
      per_page: per_page ?? 10,
    };

    return { items, pagination };
  }

  async find(id: number) {
    return this.repository.find_by_id(id);
  }

  async edit(
    product_id: number,
    { detail, seller_id, brand_id, price, categories, ...product }: ProductInputDTO,
  ) {
    try {
      const updated_product_entity = await this.entity_manager.transaction(async (manager) => {
        // 상품 디테일 업데이트
        await this.product_detail_repository
          .with_transaction(manager)
          .update({ ...detail, product_id });

        // 상품 가격 업데이트
        await this.product_price_repository
          .with_transaction(manager)
          .update({ ...price, product_id });

        // 상품 카테고리 업데이트
        for (const category of categories) {
          await this.product_category_repository
            .with_transaction(manager)
            .update({ ...category, product_id });
        }

        // 상품 제품 업데이트
        return this.repository.with_transaction(manager).update(
          {
            seller_id,
            brand_id,
            ...product,
          },
          product_id,
        );
      });

      if (!updated_product_entity) {
        throw new Error("상품 업데이트에 실패했습니다.");
      }

      return (({ id, name, slug, updated_at }) => ({
        id,
        name,
        slug,
        updated_at,
      }))(updated_product_entity);
    } catch (error) {
      throw new Error((error as Error).message);
    }
  }

  async remove(id: number) {
    return this.repository.delete(id);
  }
}

File: src/main/java/com/wanted_preonboarding_challenge_backend/eCommerce/service/ProductService.java
package com.wanted_preonboarding_challenge_backend.eCommerce.service;


import com.wanted_preonboarding_challenge_backend.eCommerce.common.PaginationDto;
import com.wanted_preonboarding_challenge_backend.eCommerce.domain.*;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.Image.request.ImageAddRequest;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.Image.response.ImageAddResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.ProductSummaryDto;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.option.request.ProductOptionAddRequest;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.option.request.ProductOptionUpdateRequest;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.option.response.ProductOptionAddResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.option.response.ProductOptionUpdateResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.product.request.*;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.product.response.ProductCreateResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.product.response.ProductDetailResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.product.response.ProductListResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.product.response.ProductUpdateResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.mapper.*;
import com.wanted_preonboarding_challenge_backend.eCommerce.respository.*;
import com.wanted_preonboarding_challenge_backend.eCommerce.util.JsonConverter;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Service
@AllArgsConstructor
public class ProductService {
    private final ProductCategoryRepository productCategoryRepository;
    private final ProductRepository productRepository;
    private final CategoryRepository categoryRepository;
    private final ProductOptionGroupRepository productOptionGroupRepository;
    private final ProductOptionRepository productOptionRepository;
    private final ProductImageRepository productImageRepository;
    private final ProductDetailRepository productDetailRepository;
    private final ProductPriceRepository productPriceRepository;
    private final SellerRepository sellerRepository;
    private final BrandRepository brandRepository;
    private final TagRepository tagRepository;
    private final ProductTagRepository productTagRepository;

    private final ProductMapper productMapper;
    private final ProductOptionGroupMapper productOptionGroupMapper;
    private final ProductOptionMapper productOptionMapper;
    private final ProductPriceMapper productPriceMapper;
    private final ProductDetailMapper productDetailMapper;
    private final ProductImageMapper productImageMapper;
    private final SellerMapper sellerMapper;
    private final BrandMapper brandMapper;
    private final PriceMapper priceMapper;
    private final DetailMapper detailMapper;


    public ProductListResponse getProductList(ProductsSearchCondition condition) {
        Pageable pageable = PageRequest.of(
                condition.getPage() - 1,
                condition.getPerPage(),
                parseSort(condition.getSort())
        );

        Page<ProductSummaryDto> page = productRepository.findByCondition(condition, pageable);

        List<ProductSummaryDto> items = page.getContent();

        return ProductListResponse.builder()
                .items(items)
                .pagination(PaginationDto.builder()
                        .totalItems(page.getTotalElements())
                        .totalPages(page.getTotalPages())
                        .currentPage(condition.getPage())
                        .perPage(condition.getPerPage())
                        .build())
                .build();
    }



    public ProductDetailResponse getProduct(Long id){
        Product product = productRepository.findWithSellerBrandPriceDetailById(id);

        List<ProductCategory> categories = productCategoryRepository.findByProductId(id);
        List<ProductOptionGroup> optionGroups = productOptionGroupRepository.findByProductId(id);
        List<ProductImage> images = productImageRepository.findByProductId(id);
        List<ProductTag> tags = productTagRepository.findByProductId(id);

        ProductDetailResponse response = ProductDetailResponse.builder()
                .id(product.getId())
                .name(product.getName())
                .shortDescription(product.getShortDescription())
                .seller(sellerMapper.toSellerDetailDto(product.getSeller()))
                .brand(brandMapper.toBrandDetailDto(product.getBrand()))
                .price(priceMapper.toPriceDto(product.getPrice()))
                .detail(detailMapper.toDetailDto(product.getDetail()))
//                .categories(categoryMapper.toDtoList(categories))
//                .optionGroups(optionGroupMapper.toDtoList(optionGroups))
//                .images(imageMapper.toDtoList(images))
//                .tags(tagMapper.toDtoList(tags))
//                .rating()
                .createdAt(product.getCreatedAt())
                .updatedAt(product.getUpdatedAt())
                .build();

        return response;

    }

    public ProductListResponse getProductListByCategoryId(Long categoryId, ProductsByCategoryCondition condition) {
        Pageable pageable = PageRequest.of(
                condition.getPage() - 1,
                condition.getPerPage(),
                parseSort(condition.getSort())
        );

        Page<ProductSummaryDto> page = productRepository.findByCondition(condition, pageable);

        List<ProductSummaryDto> items = page.getContent();

        return ProductListResponse.builder()
                .items(items)
                .pagination(PaginationDto.builder()
                        .totalItems(page.getTotalElements())
                        .totalPages(page.getTotalPages())
                        .currentPage(condition.getPage())
                        .perPage(condition.getPerPage())
                        .build())
                .build();
    }

    private Sort parseSort(String sort) {
        if (sort == null || !sort.contains(":")) {
            return Sort.by(Sort.Direction.DESC, "createdAt");
        }

        String[] parts = sort.split(":");
        String property = parts[0];
        Sort.Direction direction = parts[1].equalsIgnoreCase("asc") ? Sort.Direction.ASC : Sort.Direction.DESC;

        return Sort.by(direction, property);
    }



    public ProductCreateResponse writeProduct(ProductCreateRequest request) {
        //valid 추가

        Product product = addProduct(request);

        // product detail
        addProductDetail(request.getDetail(), product);
        // product price
        addProductPrice(request.getPrice(), product);

        // productCategory
        addProductCategory(request, product);

        // optiongroup + option
        addOptionGroup(request, product);

        // image
        addProductImage(request.getImages(), product);

        //product tag
        addProductTag(request, product);

        return productMapper.toProductCreateResponse(product);
    }

    private Product addProduct(ProductCreateRequest request) {
        // seller
        Seller seller = sellerRepository.findById(request.getSellerId()).orElseThrow(() -> new RuntimeException());
        // brand
        Brand brand = brandRepository.findById(request.getBrandId()).orElseThrow(() -> new RuntimeException());

        Product product = productMapper.toEntity(request);

        product.addSeller(seller);
        product.addBrand(brand);

        return productRepository.save(product);
    }

    private void addOptionGroup(ProductCreateRequest request, Product product) {

        request.getOptionGroups().forEach(productOptionGroupCreateRequest -> {

            ProductOptionGroup productOptionGroup =
                    productOptionGroupMapper.toEntity(productOptionGroupCreateRequest);
            productOptionGroup.addProduct(product);

            productOptionGroupRepository.save(productOptionGroup);

            productOptionGroupCreateRequest.getOptions().forEach(optionCreateRequest -> {
                ProductOption productOption =
                        productOptionMapper.toEntity(optionCreateRequest);
                productOption.addOptionGroup(productOptionGroup);

                productOptionRepository.save(productOption);
            });
        });
    }

    private void addProductDetail(ProductDetailCreateRequest request, Product product) {
        ProductDetail productDetail = productDetailMapper.toEntity(request);
        productDetail.addProduct(product);

        productDetailRepository.save(productDetail);
    }

    private void addProductPrice(ProductPriceCreateRequest request, Product product) {

        ProductPrice productPrice = productPriceMapper.toEntity(request);
        productPrice.addProduct(product);

        productPriceRepository.save(productPrice);
    }

    private void addProductImage(List<ProductImageCreateRequest> images, Product product) {
        images.forEach(image -> {
            ProductImage productImage = productImageMapper.toEntity(image);
            if (image.getOptionId() != null) {
                ProductOption option = productOptionRepository.findById(image.getOptionId())
                        .orElseThrow(() -> new IllegalArgumentException("해당 옵션이 존재하지 않습니다."));

                productImage.addProductOption(option);
            }
            productImage.addProduct(product);
            productImageRepository.save(productImage);

        });
    }

    private void addProductTag(ProductCreateRequest request, Product product) {
        List<Tag> tags = tagRepository.findAllById(request.getTags());
        Map<Long, Tag> tagMap = tags.stream()
                .collect(Collectors.toMap(Tag::getId, Function.identity()));

        request.getTags().forEach(tag -> {

            ProductTag productTag = ProductTag.builder()
                    .product(product)
                    .tag(tagMap.get(tag))
                    .build();
            productTagRepository.save(productTag);
        });
    }

    private void addProductCategory(ProductCreateRequest request, Product product) {
        List<Long> categoryIds = request.getCategories().stream().map(ProductCategoryCreateRequest::getCategoryId).collect(Collectors.toList());
        List<Category> categories = categoryRepository.findAllById(categoryIds);
        Map<Long, Category> categoryMap = categories.stream()
                .collect(Collectors.toMap(Category::getId, Function.identity()));

        request.getCategories().forEach(productCategoryCreateRequest -> {
            Long categoryId = productCategoryCreateRequest.getCategoryId();

            ProductCategory productCategory =
                    ProductCategory.builder()
                            .product(product)
                            .category(categoryMap.get(categoryId))
                            .isPrimary(productCategoryCreateRequest.getIsPrimary())
                            .build();

            productCategoryRepository.save(productCategory);
        });
    }

    public ProductUpdateResponse updateProduct(Long id, ProductCreateRequest request) {
        Product product = productRepository.findWithSellerBrandPriceDetailById(id);

        Seller seller = sellerRepository.findById(request.getSellerId()).orElseThrow(() -> new RuntimeException());
        Brand brand = brandRepository.findById(request.getBrandId()).orElseThrow(() -> new RuntimeException());

        product.update(request.getName(), request.getSlug(), request.getShortDescription(), request.getFullDescription(), request.getStatus());
        product.updateSeller(seller);
        product.updateBrand(brand);

        // toOne (detail, price) 정보 변경하기
        product.getPrice().update(
                request.getPrice().getBasePrice().doubleValue(),
                request.getPrice().getSalePrice().doubleValue(),
                request.getPrice().getCostPrice().doubleValue(),
                request.getPrice().getCurrency(),
                request.getPrice().getTaxRate().doubleValue()
        );

        product.getDetail().update(
                request.getDetail().getWeight(),
                JsonConverter.toJson(request.getDetail().getDimensions()),
                request.getDetail().getMaterials(),
                request.getDetail().getCountryOfOrigin(),
                request.getDetail().getWarrantyInfo(),
                request.getDetail().getCareInstructions(),
                JsonConverter.toJson(request.getDetail().getAdditionalInfo())
        );

        // toMany
        // 1. productCategory
        productCategoryRepository.deleteByProductId(product.getId());
        addProductCategory(request, product);

        // 2.productTag
        productTagRepository.deleteByProductId(product.getId());
        addProductTag(request, product);

        // 3.productOptionGroup + option + image
        productOptionGroupRepository.deleteByProductId(product.getId());

        // 4.productImage
        addProductImage(request.getImages(), product);

        return productMapper.toProductUpdateResponse(product);

    }

    public void deleteProduct(Long id) {
        Product product = productRepository.findById(id).orElseThrow(() -> new IllegalArgumentException("옵션 그룹이 존재하지 않습니다."));
        product.deleteProduct();
    }

    public ProductOptionAddResponse addProductOption(Long id, ProductOptionAddRequest request){
        Product product = productRepository.findById(id).orElseThrow(() -> new IllegalArgumentException("옵션 그룹이 존재하지 않습니다."));
        ProductOptionGroup productOptionGroup = productOptionGroupRepository.findById(request.getOptionGroupId()).orElseThrow(() -> new IllegalArgumentException("옵션 그룹이 존재하지 않습니다."));

        ProductOption productOption = productOptionMapper.toEntity(request);

        productOptionGroup.addOption(productOption);
        productOptionRepository.save(productOption);

        return productOptionMapper.toProductOptionAddResponse(productOption);

    }


    public ProductOptionUpdateResponse updateProductOption(Long id, Long optionId, ProductOptionUpdateRequest optionUpdateRequest) {
        Product product = productRepository.findById(id).orElseThrow(() -> new IllegalArgumentException("옵션 그룹이 존재하지 않습니다."));
        ProductOption productOption = productOptionRepository.findById(optionId).orElseThrow(() -> new IllegalArgumentException("옵션 그룹이 존재하지 않습니다."));

        productOption.updateProductOption(
                optionUpdateRequest.getName(),
                optionUpdateRequest.getAdditionalPrice().doubleValue(),
                optionUpdateRequest.getSku(),
                optionUpdateRequest.getStock(),
                optionUpdateRequest.getDisplayOrder()
        );

        return productOptionMapper.toProductOptionUpdateResponse(productOption);
    }

    public ImageAddResponse addImage(Long id, ImageAddRequest request){
        Product product = productRepository.findById(id).orElseThrow(() -> new IllegalArgumentException("옵션 그룹이 존재하지 않습니다."));
        ProductOption productOption = productOptionRepository.findById(request.getOptionId()).orElseThrow(() -> new IllegalArgumentException("옵션 그룹이 존재하지 않습니다."));

        ProductImage productImage = productImageMapper.toEntity(request);

        productImage.addProductOption(productOption);
        productImageRepository.save(productImage);

        return productImageMapper.toImageAddResponse(productImage);
    }
}
